### PR TITLE
Add ota demo over BLE transport interface

### PR DESCRIPTION
**Pull request by ravibhagavandas**
Add ota demo over BLE transport interface

Description
-----------
* Adds OTA demo over BLE transport interface in `demos/ble` folder.
* Add ota pal for Nordic port (from PR#3255)

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: demos/ble/ota_ble/ota_demo_mqtt_ble_transport.c [hunk 1]
```diff
@@ -0,0 +1,2004 @@
...
+
+
+static BaseType_t prvConnectToMQTTBroker( void )
+{
+    BaseType_t xStatus = pdFAIL;
+
+    xNetworkContextMqtt.pParams = &xBleTransportParams;
+
+    /* Attempt to connect to the MQTT broker. If connection fails, retry after
+     * a timeout. Timeout value will be exponentially increased till the maximum
```
Comment (divekarshubham) [inline demos/ble/ota_ble/ota_demo_mqtt_ble_transport.c:1299]: I believe with the current implementation this comment is incorrect.

File: demos/ble/ota_ble/ota_demo_mqtt_ble_transport.c [hunk 2]
```diff
@@ -0,0 +1,2004 @@
...
+ * establishing the MQTT connection.
+ * @param[in] pNetworkCredentialInfo Passed to the MQTT connect function when
+ * establishing the MQTT connection.
+ * @param[in] pxNetworkInterface The network interface to use for this demo.
+ *
+ * @return `EXIT_SUCCESS` if the demo completes successfully; `EXIT_FAILURE`
+ * otherwise.
+ *
+ */
+int RunOTAMQTTBLETransportDemo( void )
```
Comment (divekarshubham) [inline demos/ble/ota_ble/ota_demo_mqtt_ble_transport.c:1916]:
```suggestion
int RunOtaMqttBLETransportDemo( void )
```

File: demos/ble/ota_ble/ota_demo_mqtt_ble_transport.c [hunk 3]
```diff
@@ -0,0 +1,2004 @@
...
+/**
+ * @brief Entry point of demo.
+ *
+ * This examples demonstrates performing over the updates using MQTT packets over
+ * BLE using a companion mobile app as the proxy to the cloud. Demo creates a transport
+ * interface connection over BLE and runs the OTA and MQTT agent in separate tasks.
+ * MQTT agent is used to handle thread safety of the MQTT library and underlying transport
+ * interface.
+ *
+ * @param[in] xAwsIotMqttMode Specify if this demo is running with the AWS IoT
```
Comment (divekarshubham) [inline demos/ble/ota_ble/ota_demo_mqtt_ble_transport.c:1903]: Looks like these parameters are no longer necessary?

**Review by divekarshubham — COMMENTED** [review]
Missing `ota_config.h` and `aws_test_ota_config.h`  in `aws_test`. Please pull in the latest from https://github.com/aws/amazon-freertos/pull/3255/commits/9f865d592697dd4d376420a49be3bc37e6cbb7fe


There are references to the code signing certificate in `projects/nordic/nrf52840-dk/ses/aws_demos/bootloader/bootloader_code/crypto.c` Could you transfer the `aws_ota_code_signer_certificate.h` to the projects folder and use it there?

File: demos/ble/ota_ble/ota_demo_mqtt_ble_transport.c [hunk 4]
```diff
@@ -0,0 +1,2004 @@
...
+        {
+            ( void ) IotBleDataTransfer_SetCallback( pTransportContext->pParams->pChannel, prvBLEChannelCallback, NULL );
+
+            /* Open is a handshake proceture, so we need to wait until it is ready to use. */
+            if( xSemaphoreTake( xChannelSemaphore, pdMS_TO_TICKS( IOT_BLE_MQTT_CREATE_CONNECTION_WAIT_MS ) ) == pdTRUE )
+            {
+                LogInfo( ( "The channel was initialized successfully" ) );
+                xStatus = pdTRUE;
+            }
+            else
```
[the remark this answers is not in the document]
Comment (ravibhagavandas) [reply demos/ble/ota_ble/ota_demo_mqtt_ble_transport.c:1258]: The retry is done from the smartphone application so we don't need retry here.

File: vendors/nordic/boards/nrf52840-dk/aws_demos/bootloader/crypto.c [hunk 5]
```diff
@@ -7,6 +7,8 @@
 #include "nrf_bootloader.h"
 #include "mbedtls/base64.h"
 
+#include "aws_ota_code_signer_certificate.h"
```
[the remark this answers is not in the document]
Comment (ravibhagavandas) [reply vendors/nordic/boards/nrf52840-dk/aws_demos/bootloader/crypto.c:10]: The file cannot be reused as including the config file path will  include path for other files like `sdk_config.h`

File: vendors/nordic/boards/nrf52840-dk/aws_tests/bootloader/crypto.c [hunk 6]
```diff
@@ -7,6 +7,8 @@
 #include "nrf_bootloader.h"
 #include "mbedtls/base64.h"
 
+#include "aws_ota_code_signer_certificate.h"
```
[the remark this answers is not in the document]
Comment (ravibhagavandas) [reply vendors/nordic/boards/nrf52840-dk/aws_tests/bootloader/crypto.c:10]: The file cannot be reused as including the config file path will  include path for other files like `sdk_config.h`

**Comment by ravibhagavandas** [comment]
> There are references to the code signing certificate in `projects/nordic/nrf52840-dk/ses/aws_demos/bootloader/bootloader_code/crypto.c` Could you transfer the `aws_ota_code_signer_certificate.h` to the projects folder and use it there?

That seems to be a duplicate of the code in `vendors/nordic/boards/nrf52840-dk/aws_demos/bootloader/crypto.c` so I have changed nordic project to depend on this file instead of having two copies. I will probably remove those duplicated code in a separate PR.

File: vendors/nordic/boards/nrf52840-dk/aws_demos/bootloader/crypto.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (divekarshubham) [reply vendors/nordic/boards/nrf52840-dk/aws_demos/bootloader/crypto.c:10]: It is best to have a single source of the code signing cert, I did not see any other references to this in files other than crypto.c

File: vendors/nordic/boards/nrf52840-dk/ports/ota_pal_for_aws/ota_pal.c [hunk 7]
```diff
@@ -397,6 +398,7 @@ OtaPalStatus_t otaPal_ResetDevice( OtaFileContext_t * const pFileContext )
 
 OtaPalStatus_t otaPal_ActivateNewImage( OtaFileContext_t * const pFileContext )
 {
+    otaPal_ResetDevice( pFileContext );
```
Comment (divekarshubham) [inline vendors/nordic/boards/nrf52840-dk/ports/ota_pal_for_aws/ota_pal.c:401]: Duplicate call.

**Comment by divekarshubham** [comment]
> > There are references to the code signing certificate in `projects/nordic/nrf52840-dk/ses/aws_demos/bootloader/bootloader_code/crypto.c` Could you transfer the `aws_ota_code_signer_certificate.h` to the projects folder and use it there?
> 
> That seems to be a duplicate of the code in `vendors/nordic/boards/nrf52840-dk/aws_demos/bootloader/crypto.c` so I have changed nordic project to depend on this file instead of having two copies. I will probably remove those duplicated code in a separate PR.

Since we are updating the nordic port, might be a good idea to cleanup here.

File: vendors/nordic/boards/nrf52840-dk/ports/ota_pal_for_aws/ota_pal.c (hunk 7 above)
Comment (ravibhagavandas) [reply vendors/nordic/boards/nrf52840-dk/ports/ota_pal_for_aws/ota_pal.c:401]: Fixed.

File: vendors/nordic/boards/nrf52840-dk/aws_demos/bootloader/crypto.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (ravibhagavandas) [reply vendors/nordic/boards/nrf52840-dk/aws_demos/bootloader/crypto.c]: Rearranged  project file sdk_config.h to make this work

**Review by divekarshubham — DISMISSED** [review]

**Review by divekarshubham — APPROVED** [review]

**Review by cobusve — APPROVED** [review]

**Pull request merged**